dbms/src/Common/BitpackingPrimitives.h

```diff
@@ -0,0 +1,336 @@
+                packGroup(dst + (count * width) / 8, tmp_buffer, width);
+            }
+        }
+    }
+
+    // Unpacks a block of BITPACKING_ALGORITHM_GROUP_SIZE values
+    // Assumes both src and dst to be of the correct size
+    template <typename T>
+    static void unPackBuffer(
+        unsigned char * dst,
+        const unsigned char * src,
+        size_t count,
+        UInt8 width,
+        bool skip_sign_extension = false)
+    {
+        for (size_t i = 0; i < count; i += BITPACKING_ALGORITHM_GROUP_SIZE)
+        {
+            unPackGroup<T>(dst + i * sizeof(T), src + (i * width) / 8, width, skip_sign_extension);
+        }
+    }
+
+    // Packs a block of BITPACKING_ALGORITHM_GROUP_SIZE values
+    template <typename T>
+    static void packBlock(unsigned char * dst, const T * src, UInt8 width)
+    {
+        return packGroup(dst, src, width);
+    }
+
+    // Unpacks a block of BITPACKING_ALGORITHM_GROUP_SIZE values
+    template <typename T>
+    static void unPackBlock(
+        unsigned char * dst,
+        const unsigned char * src,
+        UInt8 width,
+        bool skip_sign_extension = false)
+    {
+        return unPackGroup<T>(dst, src, width, skip_sign_extension);
+    }
+
+    // Calculates the minimum required number of bits per value that can store all values
+    template <typename T, bool is_signed = std::numeric_limits<T>::is_signed>
+    constexpr static UInt8 minimumBitWidth(T value)
+    {
+        return findMinimumBitWidth<T, is_signed, BYTE_ALIGNED>(value, value);
+    }
+
+    // Calculates the minimum required number of bits per value that can store all values
+    template <typename T, bool is_signed = std::numeric_limits<T>::is_signed>
+    constexpr static UInt8 minimumBitWidth(const T * values, size_t count)
+    {
+        return findMinimumBitWidth<T, is_signed, BYTE_ALIGNED>(values, count);
+    }
+
+    // Calculates the minimum required number of bits per value that can store all values,
+    // given a predetermined minimum and maximum value of the buffer
+    template <typename T, bool is_signed = std::numeric_limits<T>::is_signed>
+    constexpr static UInt8 minimumBitWidth(T minimum, T maximum)
+    {
+        return findMinimumBitWidth<T, is_signed, BYTE_ALIGNED>(minimum, maximum);
+    }
+
+    constexpr static size_t getRequiredSize(size_t count, UInt8 width)
+    {
+        count = roundUpToAlgorithmGroupSize(count);
+        return ((count * width) / 8);
+    }
+
+    // round up to nearest multiple of BITPACKING_ALGORITHM_GROUP_SIZE
+    template <typename T>
+    constexpr static T roundUpToAlgorithmGroupSize(T num_to_round)
+    {
+        constexpr T mask = BITPACKING_ALGORITHM_GROUP_SIZE - 1;
+        return (num_to_round + mask) & ~mask;
+    }
+
+private:
+    template <typename T, bool is_signed, bool round_to_next_byte = false>
+    constexpr static UInt8 findMinimumBitWidth(const T * values, size_t count)
+    {
```

Comment:
Could it have another name? There are woo many overloads of `findMinimumBitWidth`.

Reply:
```cpp
    // Calculates the minimum required number of bits per value that can store all values
    template <typename T, bool is_signed = std::numeric_limits<T>::is_signed>
    constexpr static UInt8 minimumBitWidth(T value)
    {
        return findMinimumBitWidth<T, is_signed, BYTE_ALIGNED>(value, value);
    }

    // Calculates the minimum required number of bits per value that can store all values
    template <typename T, bool is_signed = std::numeric_limits<T>::is_signed>
    constexpr static UInt8 minimumBitWidth(const T * values, size_t count)
    {
        return findMinimumBitWidth<T, is_signed, BYTE_ALIGNED>(values, count);
    }

    // Calculates the minimum required number of bits per value that can store all values,
    // given a predetermined minimum and maximum value of the buffer
    template <typename T, bool is_signed = std::numeric_limits<T>::is_signed>
    constexpr static UInt8 minimumBitWidth(T minimum, T maximum)
    {
        return findMinimumBitWidth<T, is_signed, BYTE_ALIGNED>(minimum, maximum);
    }
```

I think they are impossible to resolve by mistake.

Reply:
OK